### ISSUE TITLE
4-bit conversion for non-huggingface models

### ISSUE THREAD

**Issue by keunwoochoi** [bitsandbytes-foundation/bitsandbytes#821]
Hi,

Is it possible to use bitsandbytes to quantize `torch.nn.Linear` layers to 4bit weights? Unlike the 8-bit quantization, I couldn't find any guide or docs about it.

**Comment by michaelnny** [comment]
Yes it's possible, you can check my recent project, which use 4bit QLoRA to fine-tune LLaMA model, completely decoupled from hugging face.

https://github.com/michaelnny/QLoRA-LLM

**Comment by Titus-von-Koeller** [comment]
possible duplicate of #607, to be discussed internally

**Comment by TimDettmers** [comment]
You can do this by using the 4-bit quantization functions on the weights of the model. One way to do this manually is to do

pseudocode:
```
for module in model.modules():
    if isinstance(module, nn.Linear):
          cW, state = bnb.quantize_nf4(w.data)
           w.data.copy_(cW)
```

This should get you started. We might add this to the documentation at a later time.
